underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the “License”). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.gcs;
+
+import org.jets3t.service.ServiceException;
+import org.jets3t.service.impl.rest.httpclient.GoogleStorageService;
+import org.jets3t.service.model.GSObject;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for reading a file from GCS. The main purpose is to provide a faster skip method, as
+ * the underlying implementation will read and discard bytes until the number to skip has been
+ * reached. This input stream returns 0 when calling read with an empty buffer.
+ */
+@NotThreadSafe
+public final class GCSInputStream extends InputStream {
```

Comment:
how different is the implementation from the S3InputStream? if there are mostly identical, it would make sense to use a common ancestor class